proxy/src/coco.rs

```diff
@@ -0,0 +1,467 @@
+    let project_id = project::ProjectId::from_str(&id)?;
+    let project = project::Project::open(paths, &project_id)?;
+    let mut browser = match project {
+        project::Project::Git(git_project) => git_project.browser()?,
+    };
+
+    browser.commit(radicle_surf::git::Sha1::new(sha1))?;
+
+    let history = browser.get_history();
+    let commit = history.0.first();
+
+    Ok(Commit::from(commit))
+}
+
+/// TODO(xla): Add documentation.
+pub fn tags(paths: &Paths, id: &str) -> Result<Vec<Tag>, error::Error> {
+    let project_id = project::ProjectId::from_str(&id)?;
+    let project = project::Project::open(&paths, &project_id)?;
+    let browser = match project {
+        project::Project::Git(git_project) => git_project.browser()?,
+    };
+
+    let mut tag_names = browser.list_tags()?;
+    tag_names.sort();
+
+    let mut tags: Vec<Tag> = tag_names
+        .into_iter()
+        .map(|tag_name| Tag(tag_name.name()))
+        .collect();
+
+    tags.sort();
+
+    Ok(tags)
+}
+
+/// TODO(xla): Add documentation.
+pub fn tree(paths: &Paths, id: &str, revision: &str, prefix: &str) -> Result<Tree, error::Error> {
+    let project_id = project::ProjectId::from_str(&id)?;
+    let project = project::Project::open(&paths, &project_id)?;
+
+    let mut browser = match project {
+        project::Project::Git(git_project) => git_project.browser()?,
+    };
+
+    if let Err(_err) = browser
+        .branch(surf::git::BranchName::new(&revision))
+        .or(browser.commit(surf::git::Sha1::new(&revision)))
+        .or(browser.tag(surf::git::TagName::new(&revision)))
+    {
+        return Err(error::Error::Git(surf::git::error::Error::NotBranch));
+    };
+
+    let mut path = if prefix == "/" || prefix == "" {
+        surf::file_system::Path::root()
+    } else {
+        surf::file_system::Path::from_str(&prefix)?
+    };
+
+    let root_dir = browser.get_directory()?;
+    let prefix_dir = if path.is_root() {
+        root_dir
+    } else {
+        root_dir.find_directory(&path).ok_or_else(|| {
+            radicle_surf::file_system::error::Error::Path(
+                radicle_surf::file_system::error::Path::Empty,
+            )
+        })?
+    };
+    let mut prefix_contents = prefix_dir.list_directory();
+    prefix_contents.sort();
+
+    let entries_results: Result<Vec<TreeEntry>, error::Error> = prefix_contents
+        .iter()
+        .map(|(label, system_type)| {
+            let mut entry_path = if path.is_root() {
+                let label_path =
+                    nonempty::NonEmpty::from_slice(&[label.clone()]).ok_or_else(|| {
+                        radicle_surf::file_system::error::Error::Label(
+                            radicle_surf::file_system::error::Label::Empty,
+                        )
+                    })?;
+                surf::file_system::Path(label_path)
+            } else {
+                let mut p = path.clone();
+                p.push(label.clone());
+                p
```

Comment:
I'd expect you to be able to do `path.join(label.clone())` here, if `path :: Path`

Reply:
I think you expect this to be a rust std path, but it's from surf.